### PR TITLE
More robust logic for xlsx sheet lookup; fixes #104, relates to #80

### LINKED ISSUE #104: Wrong worksheet retrieved from xlsx file exported from Google Sheets
When I use `read_excel()` on an `xlsx` file obtained by downloading a Google Sheet as `xlsx`, the `sheet =` argument doesn't behave as expected. I don't get the sheet I request, regardless of whether I provide an integer or sheet name. If I open the file in Excel, re-save and try again, everything works fine. I've confirmed this with `xlxs` files from two different Google Sheets. I don't know if Google Sheets exports weird/broken `xlsx` or what?

Here's a complete example:

https://gist.github.com/jennybc/279ab4675f5f4d275350#file-2015-05-23_read-excel-sheet-md

### LINKED ISSUE #116: Error reading sheet if position is after a chart
Hello,
read_excel fails to read a given "sheet" worksheet if its position in the workbook is after a "chart" worksheet.

Here is a link to the example file;
https://drive.google.com/open?id=0B4_ubpPkreXXY0RMeC00RlM4a0E

The example excel file (.xlsx) contains 3 worksheets;
1) Sheet1
2) Chart1
3) Sheet2 - exact copy of Sheet1
Sheet1 imports great but Sheet2 gives the error below.  I have tested on a few different Excel files and the sheet position appears to be the issue. 

e.g. output 
read_excel(path = example_excel_file, sheet = "Sheet1")
  a  b     c
1 1 10 hello
2 2 20 there
3 3 30 world

> read_excel(path = example_excel_file, sheet = "Sheet2")
> Error: Couldn't find 'xl/worksheets/sheet3.xml' in

Issue location:
It looks like this issue can be traced to the standardise_sheet() function in the read_xlsx() function.

1) when sheet = "Sheet1" (read_excel()), standardise_sheet() computes 0 and imports fine.

2) when sheet = "Sheet2", standardise_sheet() computes 2 and the import fails.  If I manually change the result of standardise_sheet() to 1 then the import works.  

standardise_sheet() needs to remove the chart objects from "sheet_names" before evaluating the match() function.

 Thank you!

### LINKED ISSUE #168: When Excel sheet tab names are numbers, read_excel finds those instead of using sheet as position
read_excel appears to be defaulting to looking for the sheet name but this causes a problem when sheet names are numbers themselves and I am actually asking Excel to use Sheet= as the position.

### LINKED ISSUE #200: wrong worksheet is read if chart objects are present
Not sure how to make this reproducible, I can DM the excel sheet if desired but don't feel comfortable posting it publicly. 

This seems related to skipping worksheets that are chart objects.

```
path <- "2016-07-27_Site Wide Transducer Data_ID.xlsm"
sheet_names <- readxl::excel_sheets(path)
sheet_names[1:5]
# [1] "Well IDs"      "Barotroll-C"   "Barotroll"     "MW-Morris1A-C" "MW-Morris1A"  
readxl::read_excel(path = path, sheet = "Barotroll")) # i.e., sheetnames[3]
# should read in "Barotroll" but instead reads in "MW-Morris1A", i.e., sheetnames[5]
# sheets 2 and 4 are chart objects
```

### PR DESCRIPTION
**Pull request by jennybc**
Current lookup assumes that xml for sheet i is in `xl/worksheets/sheeti.xml`. But it's not that simple. Instead we need to consult the i-th `sheet` node in `xl/workbook.xml` to learn the sheet's `Id`, e.g. `rId5`. Then lookup the target xml file based on this `Id` in `xl/_rels/workbook.xml.rels`.

When does this matter? Various non-Excel tools write xlsx affected by this. And it even happens in xlsx written by Excel if there are embedded chartsheets.

Thx @jimhester for writing most of this while we worked together last week.

Real example of `i` vs worksheet name vs `Id` vs xml target file (reflects test workbook sheet-xml-lookup.xlsx):

``` r
## i      name    Id                   Target
##      <chr> <chr>                    <chr>
## 1   Africa  rId3 xl/worksheets/sheet4.xml
## 2 Americas  rId4 xl/worksheets/sheet3.xml
## 3     Asia  rId5 xl/worksheets/sheet5.xml
## 4   Europe  rId6 xl/worksheets/sheet1.xml
## 5  Oceania  rId7 xl/worksheets/sheet2.xml
```

Browse the xml for this sheet [mini-gap.xlsx](https://jennybc.github.io/excelgesis/mini-gap/index.html).

### CODE REVIEW

File: src/XlsxWorkBook.h [hunk 1]
```diff
@@ -9,56 +9,126 @@
 #include "zip.h"
 
 class XlsxWorkBook {
+
+  class SheetRelations {
```
Comment (hadley) [inline src/XlsxWorkBook.h]: Might be useful to add a brief comment here, summarising the purpose of this class

File: src/XlsxWorkBook.h [hunk 2]
```diff
@@ -9,56 +9,126 @@
...
+      }
+
+      if (i != n_) {
+        names_ = Rf_lengthgets(names_, i);
+        id_ = Rf_lengthgets(id_, i);
+        n_ = i;
+      }
+    }
+
+    void parse_workbook_rels(const std::string path) {
```
Comment (hadley) [inline src/XlsxWorkBook.h]: Missing `&`

File: tests/testthat/test-sheets.R [hunk 3]
```diff
@@ -9,3 +9,47 @@ test_that("excel_sheets returns utf-8 encoded text", {
   # maybe libxls isn't reencording? It's a bit suspicious that mu is
   # \u00b5 and libxls is giving \xb5
 })
+
+test_that("informative error when requesting non-existent sheet by name", {
+  expect_error(read_excel("iris-excel.xlsx", sheet = "tulip"),
```
Comment (hadley) [inline tests/testthat/test-sheets.R]: Can you please indent long lines like:

```R
foo(
  a,
  b
)
```

File: tests/testthat/test-sheets.R [hunk 4]
```diff
@@ -9,3 +9,47 @@ test_that("excel_sheets returns utf-8 encoded text", {
...
+test_that("invalid sheet values caught", {
+  expect_error(read_excel("iris-excel.xlsx", sheet = 0),
+               "`sheet` must be positive")
+  expect_error(read_excel("iris-excel.xlsx", sheet = rep(1L, 2)),
+               "`sheet` must have length 1")
+})
+
+## #104, #168, some of #80
+## #116, #200 (chartsheet case)
+test_that("sheet data xml target is explicitly looked up [xlsx]", {
```
Comment (hadley) [inline tests/testthat/test-sheets.R]: We have a mild convention to put issue numbers in the test name: "looked up (#104, #168)" ...

You might also consider consolidating all those issues into one so you can just refer to a single number.

Comment (jennybc) [inline tests/testthat/test-sheets.R]: I have made #104 the main issue for this now.

File: src/XlsxWorkBook.h (hunk 1 above)
Comment (jennybc) [inline src/XlsxWorkBook.h]: done in 1fa3cae

File: src/XlsxWorkBook.h (hunk 2 above)
Comment (jennybc) [inline src/XlsxWorkBook.h]: done in f110164

File: tests/testthat/test-sheets.R (hunk 3 above)
Comment (jennybc) [inline tests/testthat/test-sheets.R]: done in 55ded75

File: tests/testthat/test-sheets.R (hunk 4 above)
Comment (jennybc) [inline tests/testthat/test-sheets.R]: done in 55ded75

**Comment by jennybc** [comment]
I removed `tests/testthat/sheet-xml-lookup.xls` (97c12a5) and addressed the other comments.

I'm going to interpret "otherwise, looks good" as permission to merge this myself.

**Pull request merged**